File: RUS/XYZ.py
import numpy as np
from scipy.linalg import eigh, eig
from scipy.special import factorial, factorial2

# def calculate_elastic_tensor_full(Vp, Vs, rho):
#     # Calculate the bulk modulus K and shear modulus mu
#     K = rho * Vp**2
#     mu = rho * Vs**2
    
#     # Calculate Lamé parameters lambda and mu
#     lam = K - (2 / 3) * mu
    
#     # Initialize the 6x6 elastic tensor in Voigt notation
#     C_voigt = np.zeros((6, 6))
    
#     # Fill the elastic tensor components using Voigt notation
#     C_voigt[0, 0] = lam + 2 * mu # C_1111
#     C_voigt[1, 1] = lam + 2 * mu # C_2222
#     C_voigt[2, 2] = lam + 2 * mu # C_3333
#     C_voigt[0, 1] = lam # C_1122
#     C_voigt[0, 2] = lam # C_1133
#     C_voigt[1, 0] = lam # C_2211
#     C_voigt[1, 2] = lam # C_2233
#     C_voigt[2, 0] = lam # C_3311
#     C_voigt[2, 1] = lam # C_3322
#     C_voigt[3, 3] = mu # C_1212
#     C_voigt[4, 4] = mu # C_2323
#     C_voigt[5, 5] = mu # C_3131
    
#     # Create a 4th-order tensor
#     C_full = np.zeros((3, 3, 3, 3))
    
#     # Fill the 4th-order tensor based on the Voigt notation
#     for i in range(3):
#         for j in range(3):
#             for k in range(3):
#                 for l in range(3):
#                     C_full[i, j, k, l] = C_voigt[i, j] if k == l else 0
#                     C_full[i, j, k, l] += C_voigt[k, l] if i == j else 0
#                     C_full[i, j, k, l] += C_voigt[i, k] if j == l else 0
#                     C_full[i, j, k, l] += C_voigt[j, l] if i == k else 0
    
#     return C_full

# def V_l(lam, mu, rho):
#     return np.sqrt((lam + 2*mu)/rho)

# def V_t(mu, rho):
#     return np.sqrt(mu/rho)

# lam = 0.54*1e12
# mu = 0.27*1e12 # erg/cm^3
# rho = 2.7065 # Density in g/cm^3

# # Velocities:
# Vp = V_l(lam, mu, rho) # Compressional velocity in m/s
# Vs = V_t(mu, rho) # Shear velocity in m/s

# Constants
TWOPI = 2 * np.pi
RHO = 1.0

# Elastic tensor Cijkl for the standard isotropic material
data = [
    3., 0., 0., 0., 1., 0., 0., 0., 1., 
    0., 1., 0. ,1., 0., 0., 0., 0., 0.,
    0., 0., 1., 0., 0., 0., 1., 0., 0., 
    0., 1., 0., 1., 0., 0., 0., 0., 0., 
    1., 0., 0., 0., 3., 0., 0., 0., 1.,
    0., 0., 0., 0., 0., 1., 0., 1., 0.,
    0., 0., 1., 0., 0., 0., 1., 0., 0.,
    0., 0., 0., 0., 0., 1., 0., 1., 0.,
    1., 0., 0., 0., 1., 0., 0., 0., 3          
]

# Convert to numpy array and reshape
C = np.array(data).reshape((3, 3, 3, 3), order='F')#*1e12
# Calculate the full 4th-order elastic tensor
# C = calculate_elastic_tensor_full(Vp, Vs, rho)

# Function to compute f(p, q, r) for the corner prism
def F(ip, iq, ir):
    A = B = C = 1.0
    return (A**(ip + 1) * B**(iq + 1) * C**(ir + 1) *
            factorial(ip) * factorial(iq) * factorial(ir) /
            factorial(ip + iq + ir + 3))

#def main():
# Read NN from the user
NN = int(input("PLEASE INPUT NN: "))

# Initialize arrays
MAX_SIZE = int((NN + 1)*(NN + 2)*(NN + 3)/2) # R in the original paper
GAMMA = np.zeros((MAX_SIZE, MAX_SIZE))
E = np.zeros((MAX_SIZE, MAX_SIZE))
W = np.zeros(MAX_SIZE)
FV = np.zeros(MAX_SIZE)
FW = np.zeros(MAX_SIZE)
Z = np.zeros((MAX_SIZE, MAX_SIZE))
IERR = 0

# Initialize indices and basis functions
IC = []
LB = []
MB = []
NB = []

IG = 0
for I in range(1, 4):
    for L in range(1, NN+2):
        for M in range(1, NN+2):
            for N in range(1, NN+2):
                if L + M + N > NN + 3:
                    break 
                else:
                    IG += 1
                    IC.append(I)
                    LB.append(L - 1)
                    MB.append(M - 1)
                    NB.append(N - 1)
            
NR = IG
IC = np.array(IC)
LB = np.array(LB)
MB = np.array(MB)
NB = np.array(NB)

# Compute the elements of the E and GAMMA matrices
for IG in range(NR):
    for JG in range(IG, NR):
        I = IC[IG]
        J = IC[JG]
        LS = LB[IG] + LB[JG]
        MS = MB[IG] + MB[JG]
        NS = NB[IG] + NB[JG]

        GAMMA[IG, JG] = (
            C[I-1, 0, J-1, 0] * LB[IG] * LB[JG] * F(LS-2, MS, NS) +
            C[I-1, 1, J-1, 1] * MB[IG] * MB[JG] * F(LS, MS-2, NS) +
            C[I-1, 2, J-1, 2] * NB[IG] * NB[JG] * F(LS, MS, NS-2) +
            (C[I-1, 0, J-1, 1] * LB[IG] * MB[JG] + C[I-1, 1, J-1, 0] * MB[IG] * LB[JG]) * F(LS-1, MS-1, NS) +
            (C[I-1, 0, J-1, 2] * LB[IG] * NB[JG] + C[I-1, 2, J-1, 0] * NB[IG] * LB[JG]) * F(LS-1, MS, NS-1) +
            (C[I-1, 1, J-1, 2] * MB[IG] * NB[JG] + C[I-1, 2, J-1, 1] * NB[IG] * MB[JG]) * F(LS, MS-1, NS-1)
        )
        GAMMA[JG, IG] = GAMMA[IG, JG]
        if I == J:
            E[IG, JG] = F(LS, MS, NS)
        E[JG, IG] = E[IG, JG]

# Solve the eigenvalue problem
W, Z = eigh(GAMMA, E, driver='gvd')

# Compute frequencies from eigenvalues
W = np.sqrt(np.maximum(0.0, W) / RHO) / TWOPI

# Print the lowest 36 frequencies
print(f"FREQUENCIES FOR CORNER PRISM, NN= {NN}")
for i in range(min(36, len(W))):
    print(f"{W[i]:12.5g}", end=" ")
print()

# if __name__ == "__main__":
#     main()

File: RUS/XYZ.f
! The program is named for the basis functions it uses.
      PROGRAM XYZ
! The arrays are dimensioned 252 here, sufficienf to N = 6.
      DIMENSION GAMMA(252,252),E(252,252),W(252),FV(252),FW(252),
     &C(3,3,3,3),LB(252),MB(252),NB(252),IC(252)
! The data following is the elastic tensor Cijkl for our standard
! isotropic material. Any homogeneous anisotropy can be
! described by simply changing these data to include up to 81
! different elastic constants, for a general substance in the
! presence of magnetic fields.
      DATA C/3.,3*0.,1.,3*0.,1.,0.,1.,0.,1.,7*0.,1.,3*0.,1.,3*0.,1.,
     &0.,1.,5*0.,1.,3*0.,3.,3*0.,1.,5*0.,1.,0.,1.,3*0.,1.,3*0.,1.,
     &7*0.,1.,0.,1.,0.,1.,3*0.,1.,3*0.,3 /
      DATA RHO/1./
      TWOPI=2.*ACOS(-1.)
      PRINT*,"PLEASE INPUT NN"
      READ*,NN
! The next 16 lines assign an index IG to each basis funclion (6).
      IG=0
      DO I=1,3
      DO L=1,NN+1
      DO M=1,NN+1
      DO N=1,NN+1
      IF(L+M+N.GT.NN+3) GO TO 2
      IG=IG+1
      IC(IG)=I
      LB(IG)=L-1
      MB(IG)=M-1
      NB(IG)=N-1
    2 END DO
      END DO
      END DO
      END DO
      NR=IG
! In the next 13 statements the elements of the E and Gamma
! matrices are computed.
      DO IG=1,NR
      DO JG=IG,NR
      I=IC(IG)
      J=IC(JG) ! Typo in paper, they have IG(JG), but IG is not a vector!
      LS=LB(IG)+LB(JG)
      MS=MB(IG)+MB(JG)
      NS=NB(IG)+NB(JG)
      GAMMA(IG,JG)=
     & C(I,1,J,1)*FLOAT(LB(IG)*LB(JG))*F(LS-2,MS,NS)
     & +C(I,2,J,2)*FLOAT(MB(IG)*MB(JG))*F(LS,MS-2,NS)
     & +C(I,3,J,3)*FLOAT(NB(IG)*NB(JG))*F(LS,MS,NS-2)
     & +(C(I,1,J,2)*FLOAT(LB(IG)*MB(JG))+C(I,2,J,1)*
     & FLOAT(MB(IG)*LB(JG)))*F(LS-1,MS-1,NS)
     & +(C(I,1,J,3)*FLOAT(LB(IG)*NB(JG))+C(I,3,J,1)*
     & FLOAT(NB(IG)*LB(JG)))*F(LS-1,MS,NS-1)
     & +(C(I,2,J,3)*FLOAT(MB(IG)*NB(JG))+C(I,3,J,2)*
     & FLOAT(NB(IG)*MB(JG)))*F(LS,MS-1,NS-1)
      GAMMA(JG,IG)=GAMMA(IG,JG)
      IF(I.EQ.J) E(IG,JG)=F(LS,MS,NS)   
      E(JG,IG)=E(IG,JG)  
      END DO
      END DO            
! The next line solves the eigenvalue problem (12) using the
! EISPACK13,14 subroutine RSG.
      CALL RSG(252,NR,GAMMA,E,W,0,Z,FV,FW,IERR)
      PRINT*,IERR
! Now obtain the frequencies from the eigenvalues If Cijkl is in
! 10^12dynes/cm^2, \rho is in g/cm^3, and dimensions are in cm, then
! frequencies W are in MHz
      DO I=1,NR
      W(I)=SQRT(AMAX1(0.,W(I))/RHO)/TWOPI
      END DO
! The lowest 36 frequencies are printed out: the first 6 of
! these are always zero if GAMMA is positive.
      PRINT*,"FREQUENCIES FOR CORNER PRISM, NN= ",NN
      PRINT 101,(W(I),I=1,36)
  101 FORMAT(6G12.5)
      END
! Next is a function subprogram for f(p,q,r) for the corner prism
!(Eq. 15). It is straightforward here to substitute function
! subprograms for any of the other objects we have
! considered.
      FUNCTION F(IP,IQ,IR)
      DATA A,B,C/3*1./
      F=A**(IP+1)*B**(IQ+1)*C**(IR+1)*
     &FACT(IP)*FACT(IQ)*FACT(IR)/FACT(IP+IQ+IR+3)
      RETURN
      END
! Factorial subprogram follows.
      FUNCTION FACT(N)
      FACT=1.
      IF(N.LT.2) RETURN
      DO I=2,N
      FACT=FACT*FLOAT(I) ! 1 value for closing the DO LOOP added!
      END DO
      RETURN
      END     
